pandas/core/algorithms.py

```diff
@@ -911,10 +911,15 @@ def select_n_frame(frame, columns, n, method, keep):
     if not is_list_like(columns):
         columns = [columns]
     columns = list(columns)
+    tmp = Series(frame.index)
+    frame.reset_index(inplace=True, drop=True)
     ser = getattr(frame[columns[0]], method)(n, keep=keep)
     if isinstance(ser, Series):
         ser = ser.to_frame()
```

Comment:
a way to do this might be to (for say ``nsmallest`` when multiple columns)
run nsmallest, take the last value (or values if they are duplicates), select from the frame and argsort them. Then your result are the top values from the first run + the (totalrequested - top values of the 2nd run).

e.g in your example above since nsmallest(2, 'a') yields 1 (two rows of them), you could then select the 1 from the frame and just argsort at that point.

This would allow the single column to be quite efficient and avoid the next-n problem.

I would look around and see if there is a better solution out there (IOW google and/or look at how some other stats packages do this).